design/acme-orders-challenges-crd.md

```diff
@@ -0,0 +1,388 @@
+	// Challenge will be marked as failed.
+	IssuerRef ObjectReference `json:"issuerRef"`
+}
+
+type ChallengeStatus struct {
+	Presented bool   `json:"presented"`
+	Reason    string `json:"reason"`
+	State     State  `json:"state"`
+}
+```
+
+#### Changes to existing types
+
+The scope of changes to existing types as been purposely kept as small as possible.
+
+We do however require a change to the `status` field of the Certificate resource
+type. This should not be too problematic for users, as the cert-manager controller
+should be able to tolerate this change seamlessly.
+
+```go
+// CertificateStatus defines the observed state of Certificate
+type CertificateStatus struct {
+	Conditions      []CertificateCondition `json:"conditions,omitempty"`
+    ACME            *CertificateACMEStatus `json:"acme,omitempty"`
+    /////// NEW FIELD ///////
+	LastFailureTime *metav1.Time           `json:"lastFailureTime,omitempty"`
+}
+
+// CertificateACMEStatus holds the status for an ACME issuer
+type CertificateACMEStatus struct {
+    /////// REMOVED ALL OLD FIELDS ///////
+
+    /////// NEW FIELD ///////
+	// Order contains details about the current in-progress ACME Order.
+	// If this field is not set, an Order is not in progress.
+	// This field may point to a failed or inactive Order.
+	// It is not sufficient to check for the presence of this field in order to
+	// determine whether an order is in progress.
+	// +optional
+	OrderRef *LocalObjectReference `json:"orderRef,omitempty"`
+}
+```
+
+### Controllers
+
+Two new controllers will be built to reconcile the new `Order` and `Challenge`
+types.
+
+This two controllers will take the majority of the logic that used to exist in
+the `Prepare` function of the ACME issuer.
+
+#### Order controller
+
+When an order is created, all of the spec fields must be defined.
+The Order controller will be responsible for handling the entirety of an ACME
+order.
+The 'spec' describes an order to be created. The controller will keep the 'status'
+fields up to date by periodically syncing with the ACME server.
+
+* The Order controller will attempt to create a new order with the ACME server
+if the `status.url` field is not set. It will copy details of the order back
+onto the Order resoure, to save subsequent calls to the ACME server when not required.
+
+* If the orders 'state' is 'final' (i.e. one of valid, invalid or expired) then
+the controller will take no further action and return. (TODO: should we delete challenge
+resources for this order here too?)
+
+* For each authorization on the Order, the order controller will select a challenge
+type to use to solve the authorization (based on the Issuer config and Order's
+SolverConfig) and then ensure a Challenge resource for each authorization exists.
+
+* The order controller will check the status of the Challenge resources that are
+related to it, and if any of them have changed state, it will 'resync' the order
+status with the ACME server.
```

Comment:
Does an order get resync'd if the challenges do not change state? E.g. if an order expires the authorizations may not change state. There are also occasionally bugs ([like this one](https://github.com/letsencrypt/boulder/issues/3833)) that leave an order valid but the challenges all appear pending. Would the order controller end up polling challenges forever without realizing the order is valid?

Reply:
> Does an order get resync'd if the challenges do not change state? E.g. if an order expires the authorizations may not change state.

Once an Order has a 'final' state ('Final' is defined as 'Expired', 'Invalid', 'Revoked' or 'Valid'), no more re-syncing will be done, and the Challenge resources for that Order will be deleted.

Regarding the Expired & Revoked state - because we don't resync Orders with the ACME server once they are 'valid', it is likely we'll never observe the Expired or Revoked state, instead requiring manual user intervention (note: expiry will be handled anyway, because we'll not use the certificate produced by the Order if it is nearing expiring anyway, in favour of creating a new Order).

> There are also occasionally bugs (like this one) that leave an order valid but the challenges all appear pending. Would the order controller end up polling challenges forever without realizing the order is valid?

So in these instances, it would operate okay. Because the challenge controller updates the Challenge resource with the state of the *Authorization* after calling AcceptAuthorization, that bug should not cause a problem for us.